### PR TITLE
Add docs about linearizable registers in repositories

### PR DESCRIPTION
**Pull request by DaveCTurner**
We don't mention linearizable registers in the snapshot/restore docs
today, but these things are verified for correctness by the repository
analysis API and some users with incorrect repository implementations
struggle to understand the verification errors. This commit adds some
docs to describe them and their various implementations.

### CODE REVIEW

File: docs/reference/snapshot-restore/apis/repo-analysis-api.asciidoc [hunk 1]
```diff
@@ -178,6 +178,19 @@ complete. In this case it still instructs some of the other nodes in the
...
+when accessed concurrently by multiple nodes. The detailed implementation of
+the compare-and-exchange operation on linearizable registers varies by
+repository type. Repository analysis verifies that that uncontended
+compare-and-exchange operations on a linearizable register blob always succeed,
+and that contended operations either succeed or report the contention but do
+not return incorrect results. If a contended write fails due to contention then
+{es} will keep retrying until it eventually succeeds. Most of the
+compare-and-exchange operations performed by repository analysis atomically
+increment a counter, represented as an 8-byte blob, but some operations also
+verify the behaviour on small blobs with sizes other than 8 bytes.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/repo-analysis-api.asciidoc:192]:
```suggestion
Linearizable registers are special blobs that {es} updates using a
compare-and-exchange operation. This operation ensures strong consistency, even
if the blob accessed by multiple nodes at the same time. The implementation of
the compare-and-exchange operation on linearizable registers varies by
repository type. Repository analysis verifies that that uncontended
compare-and-exchange operations on a linearizable register blob always succeed.
The analysis also verifies that contended operations either succeed or report the contention but do
not return incorrect results. If an operation fails due to contention, 
{es} retries the operation until it succeeds. Most
compare-and-exchange operations performed by repository analysis atomically
increment a counter, represented as an 8-byte blob. Some operations also
verify the behavior on small blobs with sizes other than 8 bytes.
```

File: docs/reference/snapshot-restore/repository-azure.asciidoc [hunk 2]
```diff
@@ -257,3 +257,13 @@ following naming rules:
...
+
+[[repository-azure-linearizable-registers]]
+[discrete]
+==== Linearizable register implementation
+
+The linearizable register implementation for Azure repositories is based around
+Azure's support for strongly consistent leases that can prevent concurrent
+access to a blob. To perform a compare-and-exchange operation on a register,
+{es} first obtains a lease on the blob, then uploads the updated blob under the
+same lease, which ensures that the blob has not changed in the meantime.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/repository-azure.asciidoc:269]:
```suggestion
The linearizable register implementation for Azure repositories is based on
Azure's support for strongly consistent leases. These leases can prevent concurrent
access to a blob. To perform a compare-and-exchange operation on a register,
{es} first obtains a lease on the blob, then uploads the updated blob under the
same lease. This process ensures that the blob has not changed in the meantime.
```

File: docs/reference/snapshot-restore/repository-gcs.asciidoc [hunk 3]
```diff
@@ -275,3 +275,14 @@ The service account used to access the bucket must have the "Writer" access to t
...
+[[repository-gcs-linearizable-registers]]
+[discrete]
+==== Linearizable register implementation
+
+The linearizable register implementation for GCS repositories is based around
+GCS's support for strongly consistent preconditions that can be imposed on
+put-blob operations. To perform a compare-and-exchange operation on a register,
+{es} retrieves the register blob and its current generation, and then uploads
+the updated blob with a precondition saying that the generation has not changed
+in the meantime.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/repository-gcs.asciidoc:288]:
```suggestion
The linearizable register implementation for GCS repositories is based on
GCS's support for strongly consistent preconditions. These preconditions can be imposed on
put-blob operations. To perform a compare-and-exchange operation on a register,
{es} retrieves the register blob and its current generation, and then uploads
the updated blob with a precondition. The precondition ensures that the generation has not changed
in the meantime.
```

File: docs/reference/snapshot-restore/repository-s3.asciidoc [hunk 4]
```diff
@@ -557,3 +557,13 @@ IMPORTANT: The symlink must be created on all data and master eligible nodes and
...
+
+[[repository-s3-linearizable-registers]]
+[discrete]
+==== Linearizable register implementation
+
+The linearizable register implementation for S3 repositories is based on the
+strongly consistent semantics of the multipart upload API. {es} first creates a
+multipart upload to indicate its intention to perform a linearizable register
+operation, then lists and cancels all other multipart uploads for the same
+register, and finally attempts to complete the upload.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/repository-s3.asciidoc:569]:
```suggestion
The linearizable register implementation for S3 repositories is based on the
strongly consistent semantics of the multipart upload API. {es} first creates a
multipart upload to indicate its intention to perform a linearizable register
operation. {es} then lists and cancels all other multipart uploads for the same
register. {es} then attempts to complete the upload.
```

File: docs/reference/snapshot-restore/repository-shared-file-system.asciidoc [hunk 5]
```diff
@@ -84,3 +84,13 @@ each node, but for these accounts to have different numeric user or group IDs.
...
+
+[[repository-fs-linearizable-registers]]
+[discrete]
+==== Linearizable register implementation
+
+The linearizable register implementation for shared filesystem repositories is
+based around file locking. To perform a compare-and-exchange operation on a
+register, {es} first obtains a lock on the underlying file, then writes the
+updated contents under the same lock, which ensures that the file has not
+changed in the meantime.
```
Comment (jrodewig) [inline docs/reference/snapshot-restore/repository-shared-file-system.asciidoc:96]:
```suggestion
The linearizable register implementation for shared filesystem repositories is
based around file locking. To perform a compare-and-exchange operation on a
register, {es} first locks the underlying file and then writes the
updated contents under the same lock. This ensures that the file has not
changed in the meantime.
```

**Review by jrodewig — APPROVED** [review]
LGTM. I left some suggestions that you can keep or discard as wanted. Thanks!

**Comment by DaveCTurner** [comment]
Thanks James, I mostly incorporated your suggestions

**Review by jrodewig — APPROVED** [review]

**Pull request merged**